admin-dev/themes/new-theme/js/pages/product/components/categories/categories-manager.ts

```diff
@@ -0,0 +1,152 @@
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+import ProductMap from '@pages/product/product-map';
+import ProductEventMap from '@pages/product/product-event-map';
+import CategoryTreeSelector from '@pages/product/components/categories/category-tree-selector';
+import Tags from '@pages/product/components/categories/tags';
+import {EventEmitter} from "events";
+
+const ProductCategoryMap = ProductMap.categories;
+
+export default class CategoriesManager {
+  eventEmitter: EventEmitter;
+
+  categoryTreeSelector: CategoryTreeSelector;
+
+  categoriesContainer: HTMLElement;
+
+  addCategoriesBtn: HTMLElement;
+
+  typeaheadData: Array<any>;
+
+  tags: Tags;
+
+  constructor(eventEmitter: EventEmitter) {
+    this.eventEmitter = eventEmitter;
+    this.categoryTreeSelector = new CategoryTreeSelector(eventEmitter);
+
+    const categoriesContainer = document.querySelector(ProductCategoryMap.categoriesContainer);
+
+    if (!(categoriesContainer instanceof HTMLElement)) {
+      throw '"categoriesContainer" must be a valid HTMLElement';
+    }
+    this.categoriesContainer = categoriesContainer;
+
+    const addCategoriesBtn = this.categoriesContainer.querySelector(ProductCategoryMap.addCategoriesBtn);
+
+    if (!(addCategoriesBtn instanceof HTMLElement)) {
+      throw '"addCategoriesBtn" must be a valid HTMLElement';
+    }
+
+    this.addCategoriesBtn = addCategoriesBtn;
+    this.typeaheadData = [];
+    this.tags = new Tags(
+      eventEmitter,
+      `${ProductCategoryMap.categoriesContainer} ${ProductCategoryMap.tagsContainer}`,
+    );
+    this.tags.render(this.collectCategories());
+    this.renderDefaultCategorySelection();
+    this.listenCategoryChanges();
+    this.listenDefaultCategorySelect();
+    this.initCategoryTreeModal();
+  }
+
+  private initCategoryTreeModal(): void {
+    this.addCategoriesBtn.addEventListener('click', () => this.categoryTreeSelector.showModal(
+      this.collectCategories(),
+    ));
+    this.eventEmitter.on(ProductEventMap.categories.applyCategoryTreeChanges, (eventData) => {
+      this.tags.render(eventData.categories);
+    });
+  }
+
+  private collectCategories(): Array<Category> {
+    // these are at first rendered on page load and later updated dynamically
+    const tagsContainer = this.categoriesContainer.querySelector(ProductCategoryMap.tagsContainer) as HTMLElement;
+    const tags = tagsContainer.querySelectorAll(ProductCategoryMap.tagItem);
+    const categories: Array<{ id: number; name: string; isDefault: boolean; }> = [];
```

Comment:
You could create a type for this structure, which would make the code more readable:

```suggestion
    const categories: Category[] = [];
```